src/main/java/org/apache/commons/text/similarity/LongestCommonSubsequence.java

```diff
@@ -120,24 +170,72 @@ public CharSequence longestCommonSubsequence(final CharSequence left, final Char
        if (left == null || right == null) {
            throw new IllegalArgumentException("Inputs must not be null");
        }
-       final StringBuilder longestCommonSubstringArray = new StringBuilder(Math.max(left.length(), right.length()));
-       final int[][] lcsLengthArray = longestCommonSubstringLengthArray(left, right);
-       int i = left.length() - 1;
-       int j = right.length() - 1;
-       int k = lcsLengthArray[left.length()][right.length()] - 1;
-       while (k >= 0) {
-           if (left.charAt(i) == right.charAt(j)) {
-               longestCommonSubstringArray.append(left.charAt(i));
-               i = i - 1;
-               j = j - 1;
-               k = k - 1;
-           } else if (lcsLengthArray[i + 1][j] < lcsLengthArray[i][j + 1]) {
-               i = i - 1;
-           } else {
-               j = j - 1;
-           }
+       // Find lengths of two strings
+       final int leftSz = left.length();
+       final int rightSz = right.length();
+
+       // Check if we can save even more space
+       if (leftSz < rightSz) {
+           return algorithmC(right, rightSz, left, leftSz);
        }
-       return longestCommonSubstringArray.reverse().toString();
+       return algorithmC(left, leftSz, right, rightSz);
+   }
+
+    /**
+     * An implementation of "ALG C" from Hirschberg's paper <a href="https://dl.acm.org/doi/10.1145/360825.360861">A linear space algorithm for computing maximal common subsequences</a>.
+     * Assuming the sequence <code>left</code> is of size <code>m</code> and the sequence <code>right</code> is of size <code>n</code>,
+     * this method returns Longest Common Subsequence (LCS) the two sequences.
+     * As per the paper, this method runs in O(m * n) time and O(m + n) space.
+     *
+     * @param left Left sequence
+     * @param m Length of left sequence
+     * @param right Right sequence
+     * @param n Length of right sequence
+     * @return LCS of <code>left</code> and <code>right</code>
+     */
+   static CharSequence algorithmC(final CharSequence left, final int m,
```

Comment:
@ali-ghanbari if I understood correctly, Algorithm A and Algorithm B are very similar, but I think A requires `O(mn)` time and `O(mn)` space. While Algorithm B requires `O(mn)` time and `O(m+n)` space.

And, Algorithm C is the third one in that paper, with `O(mn)` time as Algorithms A and B, and `O(m+n)` space. With the difference between B and C being that B uses temporary memory linear to `m` and `n`. Whereas Algorithm C uses a constant amount of memory.

(Past midnight here, long time with no :coffee: so forgive me if I missed something, or said something wrong above :+1: )

The `LongestCommonSubsequence` is a function supposed to be called as a `BiFunction`, with `apply()` being the main method, and entry point to the algorithm.

With the changes here, I think now `apply` uses Algorithm B. While this method with Algorithm C is called only from the public `longestCommonSubsequence`.

Is there a reason for having both implementations here, instead of just Algorithm C and calling it from the `apply` method?

Reply:
>  ... Algorithm A and Algorithm B are very similar, but I think A requires O(mn) time and O(mn) space. While Algorithm B requires O(mn) time and O(m+n) space.

Right, in terms of time complexity, yes, they all are quadratic (O(mn), to be precise) in the size of input strings.
Algorithm B has O(m+n) space complexity, while Algorithm A has O(mn) space complexity (it returns an entire DP array; just like the method that we marked as deprecated).

> And, Algorithm C is the third one in that paper, with O(mn) time...

Thanks for bring this up @kinow. The method algorithmC allocates heap space directly or indirectly (through the calls to subSequence), and there are 2m-1 calls to algorithmC. The recurrence does not seem to be easy to solve for me, but a back of the envelope calculation tells me that this implementation should not be of linear space complexity...
Looking at the space analysis proof given in the paper, it looks like the paper makes a simplifying assumption about the storage of substrings. (please correct me if I am wrong)
Perhaps that is why [https://www.ics.uci.edu/~eppstein/161/960229.html](link), emphasizes that Hirschberg's linear-spaced algorithm is used **only** for calculating length of the LCS.
Please confirm this and I can delete algorithmC and keep algorithmB only.
I can quickly adjust the rest of the class code accordingly. Thanks!

> Past midnight here, long time with ...

Thanks @kinow your constructive comments and your hard work! 👍 🥇

Reply:
Hi @ali-ghanbari !

We are almost there. Just wanted to check a few more things before digging into algorithm C and finish reading the paper and code here.

> Algorithm C calculates the LCS itself (not just its length). If the class LongestCommonSubsequence is not supposed to calculate the LCS of two strings, then there is no reason to keep Algorithm C.

Agree. Since the method to calculate the longest subsequence was public, we will have to maintain it, and will have to calculate the LCS itself. In this case, I agree we should keep the Algorithm C, since it has the best space + time.

>To implement apply method, we could use Algorithm C, as well. We could call algorithmC and take the length of the returned CharSequence and return it, but that would involve extra computations; Algorithm B can be directly used.

Right. But we would be keeping both algorithms. I think the optimization we gain from using Algorithm B and Algorithm C, is that — if I understood correctly — the `apply` method will call Algorithm B, which may use more space/memory, but won't allocate an Object for the LCS itself, instead only returning the length from the computation array/table.

On the other hand, we could keep just one algorithm, allocate the strings, and use it for both the LCS method, and for the `apply` method to return the similarity score. I **think** I prefer to keep a single algorithm for the sake of maintainability. And since C performs better, I'd keep just C.

And maybe add a note to the Javadoc that it would be possible to optimize the `apply` by using the Algorithm B of the paper (or maybe someone could come up with a way to split the Algorithm C in a way that returns the length or the LCS). But this way we would have a good improvement over the current status quo, and introduce code that is not hard to maintain.

If in the future users want to further optimize, then we have the option of adding Algorithm B back.

What do you think?

Reply:
Hello!

> We are almost there. Just wanted to check a few more things before digging into algorithm C and finish reading the paper and code here.

I just wanted to let you know that I have made some refactoring.
Instead of just passing string sequences to `algorithmB` and `algorithmC`, I have modified the methods to take the strings with their start and end index.
I also added an extra parameter to `algorithmB` so that we can traverse the string backward instead of literally reversing the input strings (in previous version of `algorithmC` we were reversing the arguments in one of the calls to `algorithmB`).
This refactoring should certainly increase the _dexterity_ of the implementation as we avoid several object creation, string reversal, and virtual method calls (a JMH might be needed anyway, as other maintainers also mentioned).
Besides that, I think this refactor was, in fact, necessary to make the implementation as similar to the original algorithm as possible: the proof for space analysis of Algorithm C says "_We assume [...] and substrings can be transferred as arguments by giving initial and final locations_."
I hope with this change, I have not interrupted your review process! Please let me know if adding more comments helps.

> I think I prefer to keep a single algorithm for the sake of maintainability. And since C performs better, I'd keep just C.
> What do you think?

I think we cannot get rid of Algorithm B, as it is being used as an auxiliary procedure in implementing Algorithm C: please see step 3 of Algorithm C in the paper.
But, I understand the maintainability concern. 
How about doing this: we could keep **only** Algorithm B and in this version to improve `apply` method only; we can keep the old implementation (dynamic programming) of the method `longestCommonSubsequence` for now.
We can mention that improving space complexity of `longestCommonSubsequence` is a work in progress.

But if we want to keep Algorithm C, we must keep Algorithm B as it is being called from Algorithm C.
And in that case it is ideal to use Algorithm B for `apply` method, because we already have the implementation and it involves less computation than Algorithm C.

Please let me know if you prefer the former approach. In that case I can revert the code for `longestCommonSubsequence`.

Thanks a lot!